src/Microsoft.NetCore.Analyzers/Core/Performance/DoNotUseCountWhenAnyCanBeUsed.cs

```diff
@@ -0,0 +1,465 @@
+            private bool IsRightCountComparison(IBinaryOperation binaryOperation)
+            {
+                if (!TryGetInt32Constant(binaryOperation.LeftOperand, out var constant))
+                {
+                    return false;
+                }
+
+                if (constant == 0 &&
+                    binaryOperation.OperatorKind != BinaryOperatorKind.Equals &&
+                    binaryOperation.OperatorKind != BinaryOperatorKind.NotEquals &&
+                    binaryOperation.OperatorKind != BinaryOperatorKind.LessThan &&
+                    binaryOperation.OperatorKind != BinaryOperatorKind.GreaterThanOrEqual)
+                {
+                    return false;
+                }
+                else if (constant == 1 &&
+                    binaryOperation.OperatorKind != BinaryOperatorKind.LessThanOrEqual &&
+                    binaryOperation.OperatorKind != BinaryOperatorKind.GreaterThan)
+                {
+                    return false;
+                }
+                else if (constant > 1)
+                {
+                    return false;
+                }
+
+                return IsCountAsyncMethodInvocationAwaited(binaryOperation.RightOperand);
+            }
+
+            /// <summary>
+            /// Tries the get an <see cref="int"/> constant from the <paramref name="operation"/>.
+            /// </summary>
+            /// <param name="operation">The operation.</param>
+            /// <param name="constant">The constant the <paramref name="operation"/> represents, if succeeded, or zero if <paramref name="operation"/> is not a constant.</param>
+            /// <returns><see langword="true" /> <paramref name="operation"/> is a constant, <see langword="false" /> otherwise.</returns>
+            private static bool TryGetInt32Constant(IOperation operation, out int constant)
+            {
+                constant = default;
+
+                if (operation?.Type?.SpecialType != SpecialType.System_Int32)
+                {
+                    return false;
+                }
+
+                var comparandValueOpt = operation.ConstantValue;
+
+                if (!comparandValueOpt.HasValue)
+                {
+                    return false;
+                }
+
+                constant = (int)comparandValueOpt.Value;
+
+                return true;
+            }
+
+            /// <summary>
+            /// Checks the <paramref name="operation"/> is an invocation of the method <see cref="targetMethod" /> in the <see cref="targetType" />.
+            /// </summary>
+            /// <param name="operation">The operation.</param>
+            /// <returns><see langword="true" /> if the <paramref name="operation"/> is an invocation of the method <see cref="targetMethod" /> in the <see cref="targetType" />; 
+            /// <see langword="false" /> otherwise.</returns>
+            private bool IsCountAsyncMethodInvocationAwaited(IOperation operation)
```

Comment:
This should be probably renamed to avoid implying it only deals with `CountAsync`, as it actually seems to deal with all possible target methods.